src/parser/on_demand/parse_to_structure.ts

```diff
@@ -0,0 +1,143 @@
+import { type Code } from '../../code';
+import { type BSONElement, getSize, parseToElements } from './parse_to_elements';
+
+/** @internal */
+const DEFAULT_REVIVER: BSONReviver = (
+  _bytes: Uint8Array,
+  _container: Container,
+  _element: BSONElement
+) => null;
+
+/** @internal */
+function parseToElementsToArray(bytes: Uint8Array, offset?: number | null): BSONElement[] {
+  const res = parseToElements(bytes, offset);
+  return Array.isArray(res) ? res : [...res];
+}
+
+/** @internal */
+type ParseContext = {
+  elementOffset: number;
+  elements: BSONElement[];
+  container: Container;
+  previous: ParseContext | null;
+};
+
+/**
+ * @experimental
+ * @public
+ * A union of the possible containers for BSON elements.
+ *
+ * Depending on kind, a reviver can accurately assign a value to a name on the container.
+ */
+export type Container =
+  | {
+      dest: Record<string, unknown>;
+      kind: 'object';
+    }
+  | {
+      dest: Map<string, unknown>;
+      kind: 'map';
+    }
+  | {
+      dest: Array<unknown>;
+      kind: 'array';
+    }
+  | {
+      dest: Code;
+      kind: 'code';
+    }
+  | {
+      kind: 'custom';
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      dest: any;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      [key: string]: any;
+    };
```

Comment:
Should this be parameterized instead of relying on any?

Something like this might work:
```typescript
export type Container<TContainer = unknown> = 
  | {
      dest: Record<string, unknown>;
      kind: 'object';
    }
  | {
      dest: Map<string, unknown>;
      kind: 'map';
    }
  | {
      dest: Array<unknown>;
      kind: 'array';
    }
  | {
      dest: Code;
      kind: 'code';
    }
  | {
      kind: 'custom';
      dest: TContainer
    };
```

Reply:
I don't see where this comes in handy, the container type is used in two places, the return type of parseToStructure and the input to a reviver. A reviver can't really work on parameterized type it needs to handle any/all of these depending on what nested layer it it operating on. For the return type of `parseToStructure` you do get something better than `any` if you pass in something with a better type for `dest`

Reply:
Isn't the point that a user's reviver might result in a different output shape?  Their reviver needs to all variations but the `custom` is the final result type of the parsing

Reply:
The value returned by parseToStructure is determined by the dest field on root. The containers returned by the reviver are to control each level of nesting. There isn't a direct relation between custom and the final return type/value.